headerfs/store.go

```diff
@@ -538,7 +538,35 @@ func (h *blockHeaderStore) WriteHeaders(hdrs ...BlockHeader) error {
 		headerLocs[i] = header.toIndexEntry()
 	}
 
-	return h.addHeaders(headerLocs)
+	// Attempt to add the headers to the database. If this fails, we'll need
+	// to roll back any changes to the header file to maintain consistency.
+	// The rollback process bases on the number of header serialized. If
+	// both the initial operation and the rollback fail, we return
+	// a detailed error explaining both failures to aid in debugging.
+	if err := h.addHeaders(headerLocs); err != nil {
+		// Since the probability of failing to write to the database is
+		// very low, it is mostly worth the cost of file sync operation
+		// to make sure truncate headers does it correctly.
+		if err := h.file.Sync(); err != nil {
+			return fmt.Errorf("failed to sync block headers "+
```

Comment:
If we can't sync, then we return here and don't truncate. I guess there's not much we can do anyway as this would hint at file system issues on the machine. But I'm wondering if we should just log this error and try to truncate anyway? Same question in `filterHeaderStore.WriteHeaders()`.

Reply:
Okay, makes sense. Thanks for the explanation.